conans/model/info.py

```diff
@@ -475,3 +475,10 @@ def default_std_non_matching(self):
 
         if self.full_settings.compiler.cppstd:
             self.settings.compiler.cppstd = self.full_settings.compiler.cppstd
+
+    def shared_library_package_id(self):
+        if self.full_options.shared:
+            for dep_name in self.requires.pkg_names:
+                dep_options = self.full_options[dep_name]
+                if "shared" not in dep_options or not self.full_options[dep_name].shared:
+                    self.requires[dep_name].package_revision_mode()
```

Comment:
Why not add just the ``package_id`` instead of setting a ``package_id_mode`` that includes revisions?

(not tested)
```python
self.requires[dep_name].package_id = self.requires[dep_name].full_package_id
```

All the other components would be handled by the value defined in Conan settings, or by specific code in the ``package_id`` method.

----

If the value of ``settings.default_package_id_mode`` is ``unrelated_mode`` (I wonder if anyone is using it), should be skip this behavior?

Reply:
> All the other components would be handled by the value defined in Conan settings, or by specific code in the package_id method.

That is the problem, for example the default version is semver, and will be mapped to only the major, which seems quite wrong for a shared library embedding a static one. If we just copy the package_id, we are failing to embed and use the new functionality of minor and patches versions downstream.

Reply:
IMO the problem is not taking into account the package ID, not the semver mode considered.

If I'm using a semver approach it is because I do want to use it*, I don't want to compile my DLL again if a minor/patch version of one of my dependencies appears. But I DO want to take into account compiler options (if using optimization, activating features,...) and they are encoded into the _package_id_.

With the current implementation, I force the user to take into account both: compiler options encoded in the package ID and also the full version string.

---

*a different argument is if the default package_id_mode is the proper one or not, and if the user is aware of this default mode. I think it should be `major_mode`+`package_id`. And I think that is what originates this issue.

Reply:
I think that the shared->static package-ID mode is broken, and this opt-in is the way to make a step forward and fix it. If we only use the major+package_id in recipes in conan-center, when users take them and use them, it will produce a very weird and difficult to debug behavior: They will have in time:
- App->B1.0(shared)->A1.1(static)
- App->B1.0(shared)->A1.2(static)
- App->B1.0(shared)->A1.3(static)

I'd say this is quite a common approach. The big problem is that these users will not run in production App the bugfixes defined by A1.2 or A1.3, because the binary embedded in B1.0 was the one built with A1.1. And they will not know, because in their dependency graph they will clearly see A1.3.

I think this opt-in is a good chance to make a small step in the right direction and avoid future headaches to users adopting packages from Conan-center.

Reply:
For Conan-center I thought we were using a much more strict mode than the default one... but yes, users can be taking the recipes and building them with a different Conan configuration.

Ok, I see the problem with the Conan graph and the opt-in workaround here.